code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,656 @@
+			shield_gen.toggle_shields()
+			return
+	..()
+
+///qdels the forcefield and calls calculate regen to update the regen value accordingly
+/obj/machinery/modular_shield_generator/proc/deactivate_shields()
+	active = FALSE
+	QDEL_LIST(deployed_shields)
+	calculate_regeneration()
+
+/obj/machinery/modular_shield_generator/attackby(obj/item/W, mob/user, params)
+
+	if(default_deconstruction_screwdriver(user,"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]open",
+		"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]closed",  W))
+		return
+
+	if(default_deconstruction_crowbar(W) && !(active) && !(recovering))
+		return
+
+	if(is_wire_tool(W) && panel_open)
+		wires.interact(user)
+		return
+
+	return ..()
+
+///toggles the forcefield on and off
+/obj/machinery/modular_shield_generator/proc/toggle_shields()
+	if(active)
+		deactivate_shields()
+		return
+	if (recovering)
+		return
+	activate_shields()
+
+
+///generates the forcefield based on the given radius and calls calculate_regen to update the regen value accordingly
+/obj/machinery/modular_shield_generator/proc/activate_shields()
+	if(active) //bug or did admin call proc on already active shield gen?
+		return
+	if(radius < 0)//what the fuck are admins doing
+		radius = initial(radius)
+	active = TRUE
+
+	if(radius >= 10) //the shield is large so we are going to use the midpoint formula and clamp it to the lowest full number in order to save processing power
+		var/fradius = round(radius)
+		var/list/inside_shield = circle_range_turfs(src, fradius - 1)//in the future we might want to apply an effect to turfs inside the shield
+		var/t1 = fradius/16
+		var/dx = fradius
+		var/dy = 0
+		var/t2
+		var/list/list_of_turfs = list()
+		while(dx >= dy)
+			list_of_turfs += locate(x + dx, y + dy, z)
+			list_of_turfs += locate(x - dx, y + dy, z)
+			list_of_turfs += locate(x + dx, y - dy, z)
+			list_of_turfs += locate(x - dx, y - dy, z)
+			list_of_turfs += locate(x + dy, y + dx, z)
+			list_of_turfs += locate(x - dy, y + dx, z)
+			list_of_turfs += locate(x + dy, y - dx, z)
+			list_of_turfs += locate(x - dy, y - dx, z)
+			dy += 1
+			t1 += dy
+			t2 = t1 - dx
+			if(t2 > 0)
+				t1 = t2
+				dx -= 1
+
+		if(exterior_only)
+			for(var/turf/target_tile as anything in list_of_turfs)
+				if (!(target_tile in inside_shield) && isspaceturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
+					var/obj/structure/emergency_shield/modular/deploying_shield = new(target_tile)
+					deploying_shield.shield_generator = src
+					deployed_shields += deploying_shield
+			calculate_regeneration()
+			active_power_usage += deployed_shields.len * BASE_MACHINE_ACTIVE_CONSUMPTION * 0.1
+			return
+
+		for(var/turf/target_tile as anything in list_of_turfs)
+			if (!(target_tile in inside_shield) && isopenturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```

Comment:
```suggestion
			if (isopenturf(target_tile) && !(target_tile in inside_shield) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```

this is better because the typecheck is less expensive than the locate/in stuff, and the placement means that we don't have to waste time doing the locate/in stuff if the turf wasn't even valid to begin with. reads nicer too.

Reply:
bump

Reply:
done